proposals/004-routing-api.md

```diff
@@ -0,0 +1,248 @@
+For this reason the `RoutingContext` does not follow the builder pattern used in the `FilterContext`, but simply
+exposes methods to asynchronously send requests down a given route.
+This allows the `Router` author to make use of the `CompletionStage` API when issuing multiple requests. 
+
+```java
+interface RoutingContext {
+
+  CompletionStage<Response> sendRequest(String route, ...)
+  void sendResponse(Resonse)
+  void disconnect()
+
+}
+```
+
+
+### Configuration
+
+Routers are configured at the top level of the proxy configuration, similarly to `filterDefinitions`:
+In addition to the `name`, `type` and `config` (which serve the same purpose for `Routers` as they do for `Filters`), a `routerDefinition` also supports a `routes` property.
+The `routes` property is optional, though any given implementation may have its own particular requirements for its `routes`.
+
+```yaml
+routerDefinitions:
+  - name: my-router
+    type: MyRouter
+    config: ...
+    routes:
+      - name: foo
+        filters: 
+          - my-first-filter
+          - my-second-filter
+        cluster: my-backing-cluster
+      - name: bar
+        filters: 
+          - my-third-filter
+        router: my-other-router
+  - name: my-other-router
+    # ...
+```
+
+A route object has a `name`, an optional list of `filters` (being the names of the filters to be applied to requests/responses that traverse this route) and either a 
+`cluster` or a `router` property, which names the receiver which will handle requests after any filters have been applied. 
+Exactly one of `cluster` or `router` must be specified.
+
+Because routers can refer to other routers they form a graph. 
+
+```
+                           my-backing-cluster
+                          ^
+                         /
+                        / foo
+     requests          /
+    ---------> my-router
+                       \
+                        \ bar              
+                         \                ...
+                          v              /
+                           my-other-router --- ...
+                                         \
+                                          ...
+```
+
+All _possible_ routes through the graph can be determined statically from the proxy configuration, but the routing of any individual incoming request is determined at runtime
+can might involve multiple outgoing requests from any given router.
+Validation performed at proxy startup will reject cyclic graphs.
+This will prevent the possibility of a request getting stuck in a router loop. 
+
+In order for non-trivial router graphs to be useful, `Router` authors will need to follow the same principle of composition as `Filter` authors.
+That is, a `Router` implementation should only talk to its receivers, and not, for example, make direct connections of its own to a backing cluster.
```

Comment:
I agree that routers should not make "out of bands" requests, but this doesn't follow from the previous "in order to be useful" to me. I mean, out of bands requests _can_ be useful, but we're discouraging them for the sake of validation, error handling, etc., I suppose.

Reply:
The idea is that all communication with the cluster should be _exclusively_ through the router/filter graph. Routers and filters should not make their own TCP connections directly to brokers.
Routers and filters may originate their own requests through the FilterContext/RouterContext API. 

Will reword.